In [1]:
import pandas as pd
import numpy as np
# Set maximum columns/rows to display as None (show all columns/rpws)
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 36)
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
data = pd.read_csv('../data/EDA - SALES DATA/sales.csv')
data.to_parquet('../data/EDA - SALES DATA/sales.parquet')
df = pd.read_parquet('../data/EDA - SALES DATA/sales.parquet')

In [3]:
df.head()

,user_id,age,sex,phone_number,joined_date,country,payment_method,loyalty_program_member,loyalty_points_redeemed,loyalty_tier,tier_discount_percentage,card_discount_percentage,coupon_discount_percentage,total_discount_percentage,total_purchase,total_discount,total_purchase_after_discount,transaction_id,payment_status,payment_date,payment_time,purchased_date,purchased_time,product_category,purchase_medium,return_status,refund_amount,return_date,order_id,released_date,estimated_delivery_date,received_date,total_delivery_days,shipping_method,shipping_cost,tracking_number,customer_exp_rating
0,****980545,22,M,63**409216**,2023-02-14,Philippines,maya,1,1,2.0,5.0,NaN,5,10,1168,116.80,1051.20,***72935436,paid,2024-03-31,21:37:17,2024-03-31,21:37:17,pet supplies,online,not returned,0,NaN,***621668,2024-04-03,2024-04-07,2024-04-07,8,standard,203,***5E7HOXAW93,3
1,****906406,23,M,63**369756**,2024-04-19,Philippines,visa_c,0,0,NaN,NaN,2.0,2,4,4617,184.68,4432.32,***68415111,paid,2024-04-24,18:26:06,2024-04-24,18:26:06,music,in-store,not returned,0,NaN,***002517,2024-04-26,2024-05-03,2024-05-04,6,standard,291,***R5NLL0ARWI,1
2,****860958,19,F,63**558004**,2023-03-03,Philippines,gcash,0,0,NaN,NaN,NaN,2,2,7526,150.52,7375.48,***52520094,paid,2024-01-26,15:34:00,2024-01-26,15:34:00,health,online,not returned,0,NaN,***050582,2024-01-28,2024-01-31,2024-02-01,11,standard,74,***EOVA6EJ4QO,1
3,****845150,58,O,63**783910**,2024-04-18,Philippines,otc,0,0,NaN,NaN,NaN,2,2,1300,26.00,1274.00,***26122828,paid,2024-06-02,20:03:12,2024-06-02,20:03:12,movies,in-store,not returned,0,NaN,***110492,2024-06-09,2024-06-15,2024-06-13,8,express,82,***WKCU4XAR9B,1
4,****664897,37,O,63**600892**,2021-09-29,Philippines,visa_d,0,0,NaN,NaN,NaN,2,2,8480,169.60,8310.40,***66502984,paid,2022-07-07,22:14:04,2022-07-07,22:14:04,clothing,in-store,not returned,0,NaN,***201483,2022-07-08,2022-07-12,2022-07-09,14,standard,51,***5DEMJF3UCZ,5


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 382423 entries, 0 to 382422
Data columns (total 37 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   user_id                        382423 non-null  object 
 1   age                            382423 non-null  int64  
 2   sex                            382423 non-null  object 
 3   phone_number                   382423 non-null  object 
 4   joined_date                    382423 non-null  object 
 5   country                        382423 non-null  object 
 6   payment_method                 382423 non-null  object 
 7   loyalty_program_member         382423 non-null  int64  
 8   loyalty_points_redeemed        382423 non-null  int64  
 9   loyalty_tier                   191126 non-null  float64
 10  tier_discount_percentage       191126 non-null  float64
 11  card_discount_percentage       88589 non-null   float64
 12  coupon_discount_percentage    

In [5]:
df.describe()

,age,loyalty_program_member,loyalty_points_redeemed,loyalty_tier,tier_discount_percentage,card_discount_percentage,coupon_discount_percentage,total_discount_percentage,total_purchase,total_discount,total_purchase_after_discount,refund_amount,return_date,total_delivery_days,shipping_cost,customer_exp_rating
count,382423.000000,382423.000000,382423.000000,191126.000000,191126.000000,88589.000000,382423.000000,382423.000000,382423.000000,382423.000000,382423.000000,382423.0,0.0,382423.000000,382423.000000,382423.000000
mean,38.972808,0.499776,2.259464,2.501753,6.254419,2.993634,2.500265,6.319557,5030.485462,317.719726,4712.765736,0.0,NaN,7.491717,175.037330,3.002795
std,12.390457,0.500001,3.031826,1.118606,2.587818,1.414830,1.706888,4.258269,2871.789926,305.853472,2702.065875,0.0,NaN,3.302425,72.496094,1.414959
min,18.000000,0.000000,0.000000,1.000000,3.000000,1.000000,0.000000,0.000000,50.000000,0.000000,42.120000,0.0,NaN,-1.000000,50.000000,1.000000
25%,28.000000,0.000000,0.000000,2.000000,5.000000,2.000000,1.000000,3.000000,2543.000000,80.400000,2376.900000,0.0,NaN,5.000000,112.000000,2.000000
50%,39.000000,0.000000,0.000000,3.000000,7.000000,3.000000,2.000000,5.000000,5037.000000,226.200000,4706.300000,0.0,NaN,7.000000,175.000000,3.000000
75%,50.000000,1.000000,5.000000,4.000000,10.000000,4.000000,4.000000,9.000000,7517.000000,464.800000,7023.350000,0.0,NaN,10.000000,238.000000,4.000000
max,60.000000,1.000000,9.000000,4.000000,10.000000,5.000000,5.000000,20.000000,9999.000000,1997.000000,9999.000000,0.0,NaN,16.000000,300.000000,5.000000


### <span style="color:red">Average age is around 39 [18-60] which is good as company cater all three types of age group {young, adult, old}</span>
### <span style="color:red">average discount is 6.3 [0-20] loyalty-tier has the highest discount follow by card and coupon, company provides thier loyal customer good discount</span>
### <span style="color:red">Refunded amount is 0 maybe cuz of no return policy</span>
### <span style="color:red">delivery_days [-1 16] avg.7.5</span>
### <span style="color:red">shipping_cost [50-300] 175 avg.</span>
### <span style="color:red">rating avg. 3 [1-5]</span>
<span style="color:red">This text is red.</span>

In [6]:
df.isnull().sum()

user_id                0
age                    0
sex                    0
phone_number           0
joined_date            0
                      ..
total_delivery_days    0
shipping_method        0
shipping_cost          0
tracking_number        0
customer_exp_rating    0
Length: 37, dtype: int64

In [7]:
df.shape[0]

382423

In [8]:
df.drop('return_date',axis=1,inplace=True)

In [9]:
df.isnull().sum()

user_id                               0
age                                   0
sex                                   0
phone_number                          0
joined_date                           0
country                               0
payment_method                        0
loyalty_program_member                0
loyalty_points_redeemed               0
loyalty_tier                     191297
tier_discount_percentage         191297
card_discount_percentage         293834
coupon_discount_percentage            0
total_discount_percentage             0
total_purchase                        0
total_discount                        0
total_purchase_after_discount         0
transaction_id                        0
payment_status                        0
payment_date                          0
payment_time                          0
purchased_date                        0
purchased_time                        0
product_category                      0
purchase_medium                       0


In [10]:
df.head(20)

,user_id,age,sex,phone_number,joined_date,country,payment_method,loyalty_program_member,loyalty_points_redeemed,loyalty_tier,tier_discount_percentage,card_discount_percentage,coupon_discount_percentage,total_discount_percentage,total_purchase,total_discount,total_purchase_after_discount,transaction_id,payment_status,payment_date,payment_time,purchased_date,purchased_time,product_category,purchase_medium,return_status,refund_amount,order_id,released_date,estimated_delivery_date,received_date,total_delivery_days,shipping_method,shipping_cost,tracking_number,customer_exp_rating
0,****980545,22,M,63**409216**,2023-02-14,Philippines,maya,1,1,2.0,5.0,NaN,5,10,1168,116.80,1051.20,***72935436,paid,2024-03-31,21:37:17,2024-03-31,21:37:17,pet supplies,online,not returned,0,***621668,2024-04-03,2024-04-07,2024-04-07,8,standard,203,***5E7HOXAW93,3
1,****906406,23,M,63**369756**,2024-04-19,Philippines,visa_c,0,0,NaN,NaN,2.0,2,4,4617,184.68,4432.32,***68415111,paid,2024-04-24,18:26:06,2024-04-24,18:26:06,music,in-store,not returned,0,***002517,2024-04-26,2024-05-03,2024-05-04,6,standard,291,***R5NLL0ARWI,1
2,****860958,19,F,63**558004**,2023-03-03,Philippines,gcash,0,0,NaN,NaN,NaN,2,2,7526,150.52,7375.48,***52520094,paid,2024-01-26,15:34:00,2024-01-26,15:34:00,health,online,not returned,0,***050582,2024-01-28,2024-01-31,2024-02-01,11,standard,74,***EOVA6EJ4QO,1
3,****845150,58,O,63**783910**,2024-04-18,Philippines,otc,0,0,NaN,NaN,NaN,2,2,1300,26.00,1274.00,***26122828,paid,2024-06-02,20:03:12,2024-06-02,20:03:12,movies,in-store,not returned,0,***110492,2024-06-09,2024-06-15,2024-06-13,8,express,82,***WKCU4XAR9B,1
4,****664897,37,O,63**600892**,2021-09-29,Philippines,visa_d,0,0,NaN,NaN,NaN,2,2,8480,169.60,8310.40,***66502984,paid,2022-07-07,22:14:04,2022-07-07,22:14:04,clothing,in-store,not returned,0,***201483,2022-07-08,2022-07-12,2022-07-09,14,standard,51,***5DEMJF3UCZ,5
5,****356860,22,F,63**651983**,2022-11-24,Philippines,bank_transfer,1,1,2.0,5.0,NaN,0,5,1526,76.30,1449.70,***55229827,paid,2023-01-09,05:44:29,2023-01-09,05:44:29,home,in-store,not returned,0,***393187,2023-01-14,2023-01-19,2023-01-17,6,standard,72,***CKPEBC3YWS,2
6,****708847,26,F,63**226399**,2023-04-25,Philippines,bank_transfer,0,0,NaN,NaN,NaN,3,3,3962,118.86,3843.14,***48223431,paid,2023-10-14,03:29:01,2023-10-14,03:29:01,furniture,online,not returned,0,***521596,2023-10-19,2023-10-26,2023-10-24,8,standard,123,***AIYCWYTLI9,3
7,****248536,18,M,63**605697**,2022-11-20,Philippines,maya,1,3,3.0,7.0,NaN,4,11,3068,337.48,2730.52,***65757756,paid,2023-01-05,22:10:36,2023-01-05,22:10:36,beauty,online,not returned,0,***223967,2023-01-10,2023-01-13,2023-01-12,8,express,119,***L1RUPOU9KI,3
8,****504114,57,M,63**165373**,2024-05-19,Philippines,mastercard_c,0,0,NaN,NaN,1.0,2,3,8891,266.73,8624.27,***10921328,paid,2024-09-23,21:38:32,2024-09-23,21:38:32,video games,in-store,not returned,0,***185380,2024-09-27,2024-10-01,2024-09-28,8,standard,176,***MTJSUYOXI0,1
9,****779825,36,F,63**724473**,2024-02-15,Philippines,gcash,1,4,2.0,5.0,NaN,1,6,4921,295.26,4625.74,***63354299,paid,2024-11-12,03:10:39,2024-11-12,03:10:39,furniture,in-store,not returned,0,***772112,2024-11-19,2024-11-26,2024-11-26,10,express,138,***9SN9IFGYXI,2


In [11]:
df['joined_date'] = pd.to_datetime(df['joined_date'])
df['payment_date'] = pd.to_datetime(df['payment_date'])
df['purchased_date'] = pd.to_datetime(df['purchased_date'])
df['released_date'] = pd.to_datetime(df['released_date'])
df['estimated_delivery_date'] = pd.to_datetime(df['estimated_delivery_date'])
df['received_date'] = pd.to_datetime(df['received_date'])

In [12]:
df.loyalty_points_redeemed.value_counts()

loyalty_points_redeemed
0    209487
7     19459
6     19287
1     19226
5     19223
4     19221
2     19216
3     19209
9     19092
8     19003
Name: count, dtype: int64

In [13]:
df_loyal = df[['loyalty_points_redeemed', 'loyalty_tier', 'tier_discount_percentage']]
df_loyal

,loyalty_points_redeemed,loyalty_tier,tier_discount_percentage
0,1,2.0,5.0
1,0,NaN,NaN
2,0,NaN,NaN
3,0,NaN,NaN
4,0,NaN,NaN
...,...,...,...
382418,0,NaN,NaN
382419,0,1.0,3.0
382420,0,NaN,NaN
382421,2,3.0,7.0


In [16]:
df_loyal = df_loyal.fillna(0, axis=0)
grouped = df_loyal.groupby('loyalty_tier')

0

In [20]:
print(grouped.count())
print('*'*70)
print(grouped.sum())

              loyalty_points_redeemed  tier_discount_percentage
loyalty_tier                                                   
0.0                            191297                    191297
1.0                             47729                     47729
2.0                             47726                     47726
3.0                             47715                     47715
4.0                             47956                     47956
**********************************************************************
              loyalty_points_redeemed  tier_discount_percentage
loyalty_tier                                                   
0.0                                 0                       0.0
1.0                            216125                  143187.0
2.0                            215067                  238630.0
3.0                            216406                  334005.0
4.0                            216473                  479560.0


### <span style="color:red">we can see pepople have redeem same amount of points across tier exclude 0 but the availed discount is diffrent as higher tier gives more discount</span>
### <span style="color:red">further enhance could be like promeoting the lower tier and make it esay for new customer to join loyalty program</span>

In [24]:
print(df.card_discount_percentage.value_counts())
df.card_discount_percentage.sum()

card_discount_percentage
1.0    17842
2.0    17776
3.0    17722
5.0    17647
4.0    17602
Name: count, dtype: int64


265203.0

In [25]:
print(df.coupon_discount_percentage.value_counts())
df.coupon_discount_percentage.sum()

coupon_discount_percentage
2    64136
3    63835
0    63732
5    63684
4    63642
1    63394
Name: count, dtype: int64


956159

### <span style="color:red">coupon discount is high then card cuz many customer didn't used card </span>

In [27]:
df.product_category.value_counts()

product_category
beauty             19433
health             19322
music              19306
pet supplies       19248
books              19178
jewelry            19157
tools              19152
groceries          19142
home               19141
office supplies    19137
automotive         19122
garden             19096
electronics        19065
appliances         19059
furniture          19025
sports             18988
clothing           18984
movies             18979
toys               18970
video games        18919
Name: count, dtype: int64